### PR TITLE
vote for pr issue association rule

### PR DESCRIPTION
**Pull request by zhangyangyu**
ref #604 

Co-authored-by: @tisonkun 
Co-authored-by: @Mini256

### CODE REVIEW

File: votes/0606-pr-issue-association.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+- 1 issue with 0 PR: question, bug vanished without a reason, etc. 
+- 1 issue with 1 PR: normal case. 
+- 1 issue with n PRs: resolve a bug multiple times([example](https://github.com/pingcap/tidb/issues/24679)), or resolve a complex issue in steps([example](https://github.com/pingcap/tidb/issues/27652)). 
+- n issues with 1 PR: trade off resolving things as a whole (due to code coupling, or we can resolve multiple bugs at the common parent level)([example](https://github.com/pingcap/tidb/pull/27697)), duplicated issues. 
+- 0 issue with 1 PR: coding without a ticket([example](https://github.com/pingcap/tidb/pull/23022)). Even if the description is detailed or even if a design document is present, pull requests focus on the change set. It is about the implementation, not the question; if it is, we always mix discussion because the change set has a primary position. 
+
+This proposal tries to resolve the last case (0 issue with 1 PR), that a pull request must associate with issue(s). In this way,
+
+- Bring issues to a first-class development concept.
```
Comment (tiancaiamao) [inline votes/0606-pr-issue-association.md:19]: For bug fixes, new feature or everything?
For example, I just want to fix a typo, and that PR still need a issue?

Comment (zhangyangyu) [reply votes/0606-pr-issue-association.md:19]: At this time, yes. We are forcing the rule so there is no exception. I think such issues won't be too many. But in future, when most contributors are already acquainted with this rule and if such issues causes too much work for contributors, we could add a new rule(such as adding a `skip-issue` label like Python community) to skip this requirement.

**Comment by zhangyangyu** [comment]
also /cc @pingcap/tidb-committers @pingcap/tidb-reviewers since this affects everyone

File: votes/0606-pr-issue-association.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+- Benefit works tracking and contributors to join force in collaboration.
+- Ensure that all changes in release notes are associated with the corresponding issues.
+- Help the release manager and users understand software evolution.
+
+### Solution
+
+Issue number in PR body will become a mandatory requirement. Currently, the PR template contains one line `Issue Number: close #xxx`, but this is not mandatory. Contributors are free to leave it empty or delete the whole content. 
+
+After this proposal, the bot will check whether there is relevant issue number(s) in the PR body, by `Issue Number: ref #{issue-number}` or `Issue Number: close[sd]?|resolve[sd]?|fix(e[sd])? #{issue-number}`. `ref` means associated with an issue but does not close it. There should be no other content in the same line and the check is case-insensitive. Multiple issues should be separated by a comma, like `Issue Number: close #12341, close #12342`. For PRs trying to close issues in a different repository, contributors need to first create an issue in the same repository and use this issue to track.
```
Comment (lance6716) [inline votes/0606-pr-issue-association.md:28]: can we tolerate “close https://github.com/pingcap/repo/issue/1234”? sometimes i like copy and paste urls and this also enables cross repo reference

File: votes/0606-pr-issue-association.md (hunk 1 above)
Comment (lance6716) [reply votes/0606-pr-issue-association.md:19]: we can create some general purpose issues, such as “fix typo“, “code refactor”, “add test” to let PR refer them 🤔

File: votes/0606-pr-issue-association.md (hunk 2 above)
Comment (sunxiaoguang) [reply votes/0606-pr-issue-association.md:28]: I think this suppose to work as long as it points to a valid issue link.

File: votes/0606-pr-issue-association.md (hunk 1 above)
Comment (sunxiaoguang) [reply votes/0606-pr-issue-association.md:19]: Flink has a 'hotfix' label to make certain simple fixes. However I think the rational behind this proposal is to make sure all release versions affected by an issue are properly fixed. So if such simple fixes affects multiple versions, it's better to have an issue to track.

File: votes/0606-pr-issue-association.md (hunk 2 above)
Comment (zhangyangyu) [reply votes/0606-pr-issue-association.md:28]: Actually we follow Github's rule https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword with some tweaks. First add ref and then disallow cross repo reference. ref adds one semantic. And forcing the issue and PR in one repo helps other members like QA and release managers' work. We recommend create a new issue in the repo and use issue to track.

Comment (lance6716) [reply votes/0606-pr-issue-association.md:28]: > it helps other members like QA and release managers' work(they may need to apply labels to the issues)

if they need add labels, less issue means less manual work. if one problem is related to many repo (like need to update dependency) more redundant issues will be spawned.

Comment (zhangyangyu) [reply votes/0606-pr-issue-association.md:28]: It makes some sense. But this means you have to take all related repos into considerations as a whole and switching back and forth to track the issue. If the repos don't share the same version strategy it also makes life harder. Updating dependencies sometimes means a bugfix so may affects multiple version so a dedicated issue could help.

File: votes/0606-pr-issue-association.md (hunk 1 above)
Comment (tiancaiamao) [reply votes/0606-pr-issue-association.md:19]: Make sense

**Comment by zhangyangyu** [comment]
ping @pingcap/tidb-maintainers @pingcap/migration-maintainers

File: votes/0606-pr-issue-association.md (hunk 2 above)
Comment (sunxiaoguang) [reply votes/0606-pr-issue-association.md:28]: Talking about repos, we are consolidating repos. It's going to help reducing redundant issues.

**Review by sunxiaoguang — APPROVED** [review]

**Review by tiancaiamao — APPROVED** [review]

**Pull request merged**